### PR TITLE
Latest trade data

### LINKED ISSUE #42: support for Latest Multi Trades
The alpaca endpoint `GET/v2/stocks/trades/latest` supports getting latest trades for multiple symbols at once. This would be useful for performance and avoiding quota limits.

I'd be happy to work up a PR in similar vein to the existing last_quote mod if of interest

### PR DESCRIPTION
**Pull request by sbeam**
This uses the [Latest Multi Trades endpoint](https://alpaca.markets/docs/api-references/market-data-api/stock-pricing-data/historical/#latest-multi-trades) to get the latest trade data for one or more symbols at once. This was heavily copied from the last_quote mod, but changed to deal with the JSON response as nested objects. I elected to return a vec of Trades with the symbol attached because it suits my needs. closes #42

### CODE REVIEW

**Comment by d-e-s-o** [comment]
Nice! Thanks! I will take a closer look later this week, but one quick thought: Instead of copying the `last_quote` module to `last_trade` and then modifying it to work with multiple symbols, would you be up for modifying `last_quote` to gain multi symbol support first? In my opnion it's more universal, so I don't see a downside. Last I checked I recall that there was a similar multi symbol endpoint there.

**Comment by sbeam** [comment]
yes that would make sense, they are similar and both have a single symbol endpoint and multi. However the `last_quote` currently returns a single Quote, and changing it to a Vec would be breaking. Maybe would be best to have `last_quote`, `last_quote_multi`, `last_trade` and `last_trade_multi`? So all is nice and consistent?

**Comment by d-e-s-o** [comment]
> yes that would make sense, they are similar and both have a single symbol endpoint and multi. However the `last_quote` currently returns a single Quote, and changing it to a Vec would be breaking.

But to be honest, backwards compatibility is not much of a concern for me at this point. The API surface is too large to nail everything early on, enabling more without a need is a maintenance burden, and stuff is changing upstream. So don't worry about it, we just bump the crate's version with the next release.

> Maybe would be best to have `last_quote`, `last_quote_multi`, `last_trade` and `last_trade_multi`? So all is nice and consistent?

It's an option, yeah. From my perspective, though, one is basically a superset of the other and so we may as well opt for the fewer-lines-of-code option and just stick with the multi versions of both.

**Comment by sbeam** [comment]
got it. I'll work on this at some point this week

**Comment by sbeam** [comment]
@d-e-s-o ok, see what you think. I wanted to use this library in my project to learn Rust and it's been very helpful in that regard! To reduce the copy/paste factor, chose to implement a couple generic traits to pull the data objects out of the json -- maybe not worth the trouble but it works. Also added examples etc

File: Cargo.toml [hunk 1]
```diff
@@ -21,6 +21,15 @@ A crate for interacting with the Alpaca API.
...
 
+[[example]]
+name = "latest-quote-and-trade"
+
+[[example]]
+name = "order"
+
+[[example]]
+name = "stream-realtime-data"
+
```
Comment (d-e-s-o) [inline Cargo.toml:32]: Can you please create a separate commit for these changes?

File: src/data/v2/mod.rs [hunk 2]
```diff
@@ -8,9 +8,14 @@ mod unfold;
 pub mod bars;
 /// Functionality for retrieval of the most recent quote.
 pub mod last_quote;
+/// Functionality for retrieval of the most recent trade(s).
+pub mod latest_trade;
```
Comment (d-e-s-o) [inline src/data/v2/mod.rs:12]: Why are we calling one "last" and the other "latest"? Seems inconsistent.

File: src/data/v2/last_quote.rs [hunk 3]
```diff
@@ -173,12 +199,24 @@ mod tests {
...
+    assert!(quotes[0].time >= Utc::now() - Duration::weeks(2));
+  }
+
+  /// Retrieve multiple symbols at once.
+  #[test(tokio::test)]
+  async fn request_last_quotes_multi() {
+    let api_info = ApiInfo::from_env().unwrap();
+    let client = Client::new(api_info);
+
+    let req = LastQuoteReq::new("SPY,QQQ,MSFT");
```
Comment (d-e-s-o) [inline src/data/v2/last_quote.rs:216]: I don't think we should be passing in some kind of pre-formatted string. It should likely be a vec/iter/slice/whatever of strings. If I recall correctly there is precedent in other parts of the API surface that we should follow, unless not suitable for some reason.

**Review by d-e-s-o — COMMENTED** [review]
Sorry for the delay.

> @d-e-s-o ok, see what you think. I wanted to use this library in my project to learn Rust and it's been very helpful in that regard! To reduce the copy/paste factor, chose to implement a couple generic traits to pull the data objects out of the json -- maybe not worth the trouble but it works. Also added examples etc

Yeah, I'd say let's not add the trait. It doesn't really add value from my perspective and just complicates things. I'd also suggest we keep the pull request concentrated on the plan we laid out earlier.

File: src/data/v2/last_quote.rs (hunk 3 above)
Comment (sbeam) [reply src/data/v2/last_quote.rs:216]: it's what alpaca wants, but I agree it's a bit icky. will fix

**Comment by sbeam** [comment]
@d-e-s-o ok, updated as requested and squashed

**Comment by d-e-s-o** [comment]
> @d-e-s-o ok, updated as requested and squashed

Thank you! Will have reviewed by the end of the weekend the latest.

File: src/data/v2/last_quote.rs [hunk 4]
```diff
@@ -10,77 +10,101 @@ use serde::Deserialize;
...
   /// The data feed to use.
-  #[serde(rename = "feed")]
   pub feed: Option<Feed>,
 }
 
-
-/// A helper for initializing [`LastQuoteReq`] objects.
-#[derive(Clone, Debug, Default, Eq, PartialEq)]
-#[allow(missing_copy_implementations)]
-pub struct LastQuoteReqInit {
```
Comment (d-e-s-o) [inline src/data/v2/last_quote.rs:34]: I don't see a good reason why this type is being removed?

File: src/data/v2/last_quote.rs [hunk 5]
```diff
@@ -10,77 +10,101 @@ use serde::Deserialize;
...
 }
 
-
-/// A quote bar as returned by the /v2/stocks/<symbol>/quotes/latest endpoint.
+/// A quote bar as returned by the /v2/stocks/quotes/latest endpoint.
+/// See
+/// https://alpaca.markets/docs/api-references/market-data-api/stock-pricing-data/historical/#latest-multi-quotes
 // TODO: Not all fields are hooked up.
-#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
+#[derive(Clone, Debug, Serialize, Deserialize, Eq, PartialEq)]
```
Comment (d-e-s-o) [inline src/data/v2/last_quote.rs:47]: Why is this change necessary?

File: src/data/v2/last_quote.rs [hunk 6]
```diff
@@ -10,77 +10,101 @@ use serde::Deserialize;
...
 #[non_exhaustive]
 pub struct Quote {
   /// The time stamp of this quote.
-  #[serde(rename = "t")]
```
Comment (d-e-s-o) [inline src/data/v2/last_quote.rs:62]: Why are these renames being removed?

File: src/data/v2/last_quote.rs [hunk 7]
```diff
@@ -10,77 +10,101 @@ use serde::Deserialize;
...
+  ap: Num,
+  r#as: u64,
+  bp: Num,
+  bs: u64,
+}
+
+/// A representation of the JSON data in the response
+#[derive(Debug, Deserialize)]
+pub struct LastQuoteResponse {
+  quotes: HashMap<String, QuoteDataPoint>,
```
Comment (d-e-s-o) [inline src/data/v2/last_quote.rs:100]: Why a `HashMap` repesentation? We should likely expose what we get.

File: src/data/v2/last_quote.rs [hunk 8]
```diff
@@ -94,38 +118,23 @@ EndpointNoParse! {
     Some(DATA_BASE_URL.into())
   }
 
-  fn path(input: &Self::Input) -> Str {
-    format!("/v2/stocks/{}/quotes/latest", input.symbol).into()
+  fn path(_input: &Self::Input) -> Str {
+    format!("/v2/stocks/quotes/latest").into()
```
Comment (d-e-s-o) [inline src/data/v2/last_quote.rs:122]: Don't need to format if you are not formatting.

File: src/data/v2/last_quote.rs [hunk 9]
```diff
@@ -94,38 +118,23 @@ EndpointNoParse! {
...
-      .map_err(Self::ConversionError::from)
+    Quote::parse(body).map_err(Self::ConversionError::from)
   }
 
   fn parse_err(body: &[u8]) -> Result<Self::ApiError, Vec<u8>> {
     from_json::<Self::ApiError>(body).map_err(|_| body.to_vec())
   }
 }
 
-
```
Comment (d-e-s-o) [inline src/data/v2/last_quote.rs:128]: Keep needless useless changes to a minimum?

File: src/data/v2/last_quote.rs [hunk 10]
```diff
@@ -138,33 +147,49 @@ mod tests {
...
-      "bp": 387.67,
-      "bs": 1,
-      "c": [
-        "R"
-      ]
-}"#;
-
-    let quote = from_json::<Quote>(response).unwrap();
+			"quotes": {
+				"TSLA": {
```
Comment (d-e-s-o) [inline src/data/v2/last_quote.rs:156]: Please no tabs in code using spaces?

**Review by d-e-s-o — CHANGES_REQUESTED** [review]
I won't be able to land this code without major modifications and I don't have the time to write 2+ hour reviews for each iteration, which is pretty much guaranteed to annoy the both of us.

My suggestion is that you create a pull request for the changes to the `last_quote` module first (renaming it to `last_quotes` in the process and addressing the comments I left there). We get that in. Then we work on the rest.

I expect you to understand existing code and why it does what it does or at least accept that it is there for a reason and not change it randomly, though, please. The general guidance for open source contributions to be accepted is to mirror existing project patterns, styles, etc. I am happy to explain/fix minor things, but I don't have the time or nerves to rework changes that make no attempt at blending in or doing research on existing patterns.

**Comment by sbeam** [comment]
When you mentioned that back-compat wasn't a concern, I took that as more license to refactor than it seems you intended. I understand your points and appreciate the concern for quality and following the prior patterns.

Most of these changes were to make it work with the multiple quotes JSON, which has a different structure than the single quote response. So the way the response is deserialized and coerced into a `Vec<Quote>` had to be much different, and that was the best way I found to do it. However, we don't have to return a `Vec<Quote>`, it could be a `HashMap<String, Quote>`, which would be exactly what the JSON looks like and therefore simpler to implement.

That said, I'm most interested in getting the Last Trade part of this in. Would it be more agreeable to go back to the original plan, and copy the existing pattern into last_trade, using the single object API instead of multi? There is no reason we have to implement multi here -- it was kind of a "bonus" in my mind -- but it could be done as part of an followup later on.

**Comment by d-e-s-o** [comment]
> That said, I'm most interested in getting the Last Trade part of this in. Would it be more agreeable to go back to the original plan, and copy the existing pattern into last_trade, using the single object API instead of multi? There is no reason we have to implement multi here -- it was kind of a "bonus" in my mind -- but it could be done as part of an followup later on.

In a way I coerced you into doing the last quote multi conversion, so perhaps we can do the following: I can take a closer look at getting `last_quote` converted into `last_quotes` in the next few days. Then you can build the last trade data retrieval in a similar vein. That should eliminate some of the guess work on your end and a bunch of communication overhead on mine and you can focus on the part you have an interest in. Does that sound reasonable?

**Comment by sbeam** [comment]
Yes, whichever way you want to go is fine of course. If you don't have time to convert this to multi then I'd be happy to put up another PR that just implements last_trade (single) and doesn't touch anything else.

**Comment by d-e-s-o** [comment]
Let me take a look tomorrow and then we'll take it from there.

**Comment by d-e-s-o** [comment]
> Let me take a look tomorrow and then we'll take it from there.

Okay, it's mostly ready at https://github.com/d-e-s-o/apca/tree/topic/last-quote-multi. Will probably land that tomorrow and don't expect many mor echanges.

**Comment by d-e-s-o** [comment]
Closing this pull request as it hasn't seen any response or update.
